src/crypto/hashing/merkle.rs

```diff
@@ -81,20 +81,93 @@ impl MerkleTree {
             }
             // If there's one node left over, add it to the end
             if prev_layer.len() % 2 == 1 {
-                current_layer.push(prev_layer.len() - 1);
+                current_layer.push(prev_layer[prev_layer.len() - 1]);
             }
         }
         MerkleTree {
             nodes,
             root: current_layer[0],
-            leaves: leaves.len(),
+            size: leaves.len(),
         }
     }
+
+    /// Generate a proof that the given item is contained within the Merkle tree.
+    pub fn construct_proof(&self, index: usize) -> Vec<Hash> {
+        if index >= self.size {
+            panic!()
+        } else {
+            let mut proof = vec![];
+            let mut relative_index = index;
+            let mut current = &self.nodes[self.root];
+            while current.size > 1 {
```

Comment:
Should there be a check of index < current.size on each iteration?

Reply:
Sorry I meant relative_index not index

Reply:
We only need to check once, as long as left_child_size(x) < x for all x.